File: skills/ow-fandom-hero-data/scripts/fetch_all_hero_details.py
#!/usr/bin/env python3
"""Fetch all hero details using a hero list JSON/TXT file."""

from __future__ import annotations

import argparse
import json
import time
import urllib.parse
from pathlib import Path

from fetch_hero_details import build_hero_url, fetch_text, parse_hero_page

DEFAULT_URL_TEMPLATE = "https://overwatch.fandom.com/wiki/{heroName}"
DEFAULT_USER_AGENT = "Mozilla/5.0 (compatible; ow-fandom-hero-data/1.0)"


def normalize_hero_name(value: str) -> str:
    return value.replace("_", " ").strip()


def load_heroes(path: Path) -> list[str]:
    if not path.exists():
        raise FileNotFoundError(f"Heroes file not found: {path}")

    heroes: list[str] = []
    suffix = path.suffix.lower()

    if suffix == ".json":
        data = json.loads(path.read_text(encoding="utf-8"))
        if isinstance(data, dict) and isinstance(data.get("heroes"), list):
            source_heroes = data["heroes"]
        elif isinstance(data, list):
            source_heroes = data
        else:
            raise ValueError("JSON heroes file must be a list or an object with a 'heroes' list")

        for item in source_heroes:
            if isinstance(item, str):
                name = normalize_hero_name(item)
                if name:
                    heroes.append(name)
            elif isinstance(item, dict) and isinstance(item.get("hero"), str):
                name = normalize_hero_name(item["hero"])
                if name:
                    heroes.append(name)
    else:
        for raw_line in path.read_text(encoding="utf-8").splitlines():
            line = raw_line.strip()
            if not line or line.startswith("#"):
                continue
            heroes.append(normalize_hero_name(line))

    seen: set[str] = set()
    deduped: list[str] = []
    for hero in heroes:
        if hero and hero not in seen:
            deduped.append(hero)
            seen.add(hero)

    return deduped


def resolve_html_file(html_dir: Path, hero: str) -> Path:
    slug = hero.replace(" ", "_")
    candidates = [
        html_dir / f"{slug}.html",
        html_dir / f"{urllib.parse.quote(slug, safe='')}.html",
    ]
    for candidate in candidates:
        if candidate.exists():
            return candidate
    raise FileNotFoundError(
        f"No local HTML file found for hero '{hero}' in {html_dir} "
        f"(tried: {', '.join(str(c.name) for c in candidates)})"
    )


def write_output(payload: dict, output: Path | None, pretty: bool) -> None:
    content = json.dumps(payload, ensure_ascii=False, indent=2 if pretty else None)
    if output:
        output.parent.mkdir(parents=True, exist_ok=True)
        output.write_text(content + "\n", encoding="utf-8")
    else:
        print(content)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Fetch details for all heroes listed in a JSON/TXT file. "
            "Use fetch_heroes_from_page.py output as input."
        )
    )
    parser.add_argument(
        "--heroes-file",
        type=Path,
        required=True,
        help="Path to heroes JSON/TXT (e.g. output from fetch_heroes_from_page.py)",
    )
    parser.add_argument(
        "--html-dir",
        type=Path,
        help="Optional local HTML directory for offline parsing. File name format: <Hero_Name>.html",
    )
    parser.add_argument(
        "--url-template",
        default=DEFAULT_URL_TEMPLATE,
        help="URL template with {heroName}, {hero_name}, or {hero}",
    )
    parser.add_argument("--timeout", type=float, default=20.0, help="Network timeout in seconds")
    parser.add_argument("--user-agent", default=DEFAULT_USER_AGENT, help="HTTP User-Agent header")
    parser.add_argument("--delay", type=float, default=0.3, help="Delay seconds between hero requests")
    parser.add_argument("--max", type=int, help="Optional max number of heroes to process")
    parser.add_argument("--output", type=Path, help="Output JSON path (default: stdout)")
    parser.add_argument("--pretty", action="store_true", help="Pretty-print JSON output")
    parser.add_argument("--fail-on-error", action="store_true", help="Exit with code 1 if any hero fails")
    return parser.parse_args()


def main() -> int:
    args = parse_args()

    heroes = load_heroes(args.heroes_file)
    if args.max is not None and args.max >= 0:
        heroes = heroes[: args.max]

    results: list[dict] = []
    failures: list[dict[str, str]] = []

    for index, hero in enumerate(heroes):
        try:
            if args.html_dir:
                html_path = resolve_html_file(args.html_dir, hero)
                html = html_path.read_text(encoding="utf-8")
                source_url = str(html_path)
            else:
                source_url = build_hero_url(hero, args.url_template)
                html = fetch_text(source_url, timeout=args.timeout, user_agent=args.user_agent)

            parsed = parse_hero_page(html, hero_hint=hero)
            parsed["url"] = source_url
            results.append(parsed)
        except Exception as exc:  # pylint: disable=broad-except
            failures.append({"hero": hero, "error": str(exc)})

        is_last = index == len(heroes) - 1
        if not is_last and args.delay > 0:
            time.sleep(args.delay)

    payload = {
        "source": str(args.heroes_file),
        "count": len(results),
        "failed_count": len(failures),
        "heroes": results,
        "failures": failures,
    }
    write_output(payload, output=args.output, pretty=args.pretty)

    if args.fail_on_error and failures:
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/fandom/fetch-all-hero-details.ts
#!/usr/bin/env -S node --import tsx

import fs from "node:fs/promises";
import path from "node:path";
import {
  buildHeroUrl,
  DEFAULT_URL_TEMPLATE,
  DEFAULT_USER_AGENT,
  fetchText,
  normalizeHeroName,
  parseHeroPage,
  writeJson,
} from "../lib/fandom.js";

type Args = {
  heroesFile: string;
  htmlDir: string | null;
  urlTemplate: string;
  timeout: number;
  userAgent: string;
  delay: number;
  max: number | null;
  output: string | null;
  pretty: boolean;
  failOnError: boolean;
};

function parseArgs(argv: string[]): Args {
  const args: Args = {
    heroesFile: "",
    htmlDir: null,
    urlTemplate: DEFAULT_URL_TEMPLATE,
    timeout: 20,
    userAgent: DEFAULT_USER_AGENT,
    delay: 0.3,
    max: null,
    output: null,
    pretty: false,
    failOnError: false,
  };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index]!;
    switch (arg) {
      case "--heroes-file":
        args.heroesFile = argv[++index] ?? "";
        break;
      case "--html-dir":
        args.htmlDir = argv[++index] ?? "";
        break;
      case "--url-template":
        args.urlTemplate = argv[++index] ?? "";
        break;
      case "--timeout":
        args.timeout = Number(argv[++index] ?? "20");
        break;
      case "--user-agent":
        args.userAgent = argv[++index] ?? "";
        break;
      case "--delay":
        args.delay = Number(argv[++index] ?? "0.3");
        break;
      case "--max":
        args.max = Number(argv[++index] ?? "0");
        break;
      case "--output":
        args.output = argv[++index] ?? "";
        break;
      case "--pretty":
        args.pretty = true;
        break;
      case "--fail-on-error":
        args.failOnError = true;
        break;
      case "-h":
      case "--help":
        console.log("Usage: tools/fandom/fetch-all-hero-details.ts --heroes-file PATH [--html-dir PATH] [--output PATH]");
        process.exit(0);
      default:
        throw new Error(`Unknown argument: ${arg}`);
    }
  }
  if (!args.heroesFile) {
    throw new Error("--heroes-file is required");
  }
  return args;
}

async function loadHeroes(filePath: string): Promise<string[]> {
  const raw = await fs.readFile(filePath, "utf8");
  const suffix = path.extname(filePath).toLowerCase();
  const heroes: string[] = [];

  if (suffix === ".json") {
    const data = JSON.parse(raw) as unknown;
    const sourceHeroes =
      Array.isArray(data) ? data : typeof data === "object" && data && Array.isArray((data as { heroes?: unknown }).heroes)
        ? (data as { heroes: unknown[] }).heroes
        : null;
    if (!sourceHeroes) {
      throw new Error("JSON heroes file must be a list or an object with a 'heroes' list");
    }
    for (const item of sourceHeroes) {
      if (typeof item === "string") {
        const hero = normalizeHeroName(item);
        if (hero) {
          heroes.push(hero);
        }
      } else if (typeof item === "object" && item && typeof (item as { hero?: unknown }).hero === "string") {
        const hero = normalizeHeroName((item as { hero: string }).hero);
        if (hero) {
          heroes.push(hero);
        }
      }
    }
  } else {
    for (const rawLine of raw.split(/\r?\n/)) {
      const line = rawLine.trim();
      if (!line || line.startsWith("#")) {
        continue;
      }
      heroes.push(normalizeHeroName(line));
    }
  }

  const seen = new Set<string>();
  return heroes.filter((hero) => {
    if (!hero || seen.has(hero)) {
      return false;
    }
    seen.add(hero);
    return true;
  });
}

async function resolveHtmlFile(htmlDir: string, hero: string): Promise<string> {
  const slug = hero.replaceAll(" ", "_");
  const candidates = [path.join(htmlDir, `${slug}.html`), path.join(htmlDir, `${encodeURIComponent(slug)}.html`)];
  for (const candidate of candidates) {
    try {
      await fs.access(candidate);
      return candidate;
    } catch {
      // ignore
    }
  }
  throw new Error(`No local HTML file found for hero '${hero}' in ${htmlDir}`);
}

async function sleep(seconds: number): Promise<void> {
  await new Promise((resolve) => setTimeout(resolve, seconds * 1000));
}

async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2));
  let heroes = await loadHeroes(args.heroesFile);
  if (args.max !== null && args.max >= 0) {
    heroes = heroes.slice(0, args.max);
  }

  const results: unknown[] = [];
  const failures: Array<{ hero: string; error: string }> = [];

  for (let index = 0; index < heroes.length; index += 1) {
    const hero = heroes[index]!;
    try {
      let html: string;
      let sourceUrl: string;
      if (args.htmlDir) {
        const htmlPath = await resolveHtmlFile(args.htmlDir, hero);
        html = await fs.readFile(htmlPath, "utf8");
        sourceUrl = htmlPath;
      } else {
        sourceUrl = buildHeroUrl(hero, args.urlTemplate);
        html = await fetchText(sourceUrl, args.timeout, args.userAgent);
      }
      results.push({ ...parseHeroPage(html, hero), url: sourceUrl });
    } catch (error) {
      failures.push({ hero, error: (error as Error).message });
    }

    if (index < heroes.length - 1 && args.delay > 0) {
      await sleep(args.delay);
    }
  }

  await writeJson(
    args.output,
    {
      source: args.heroesFile,
      count: results.length,
      failed_count: failures.length,
      heroes: results,
      failures,
    },
    args.pretty,
  );

  if (args.failOnError && failures.length > 0) {
    process.exit(1);
  }
}

main().catch((error: Error) => {
  console.error(error.message);
  process.exit(1);
});
